### PR TITLE
[HIVEMALL-162] Support L1 normalization

### PR DESCRIPTION
**Pull request by takuti**
## What changes were proposed in this pull request?

Support `l1_normalize` in a similar manner to `l2_normalize`

## What type of PR is it?

Feature

## What is the Jira issue?

- https://issues.apache.org/jira/browse/HIVEMALL-59
- https://issues.apache.org/jira/browse/HIVEMALL-162

## How was this patch tested?

Unit test and manual test on EMR

## Checklist

(Please remove this section if not needed; check `x` for YES, blank for NO)

- [x] Did you apply source code formatter, i.e., `mvn formatter:format`, for your commit?
- [x] Did you run system tests on Hive (or Spark)?

### CODE REVIEW

**Comment by takuti** [comment]
Another option is to provide generic `normalize` interface like [sklearn](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.normalize.html)

**Comment by myui** [comment]
@takuti 

`normalize` may be introduced to Hive default UDF later as unicode normalization or so.
So, `l1_normalize` is preferred.

Could you add gitbook documentation about it?

**Comment by takuti** [comment]
Ah, good point. I've updated the document.

**Comment by myui** [comment]
Please link and close this issue as well. 
https://issues.apache.org/jira/browse/HIVEMALL-59

File: core/src/main/java/hivemall/ftvec/scaling/L1NormalizationUDF.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                features[i] = ft[0];
+                weights[i] = 1.f;
+                absoluteSum += 1.d;
+            } else if (ftlen == 2) {
+                features[i] = ft[0];
+                float v = Float.parseFloat(ft[1]);
+                weights[i] = v;
+                absoluteSum += Math.abs(v);
+            } else {
+                throw new IllegalArgumentException("Invalid feature value representation: " + s);
```
Comment (myui) [inline core/src/main/java/hivemall/ftvec/scaling/L1NormalizationUDF.java]: This is my bad...
https://github.com/apache/incubator-hivemall/blob/master/core/src/main/java/hivemall/ftvec/scaling/L2NormalizationUDF.java#L62

This is my bad.

HiveException (or UDFArgumentException) is expected here. RuntimeException should not be thrown.

**Comment by myui** [comment]
LGTM. Please merge it EMR testing.

**Comment by myui** [comment]
It seems recent roaring bitmap release caused a problem.
https://github.com/RoaringBitmap/RoaringBitmap/issues/197

We need to fix versions of depending libraries.
https://github.com/apache/incubator-hivemall/blob/master/core/pom.xml#L142

I'll fix it.

**Comment by myui** [comment]
https://github.com/apache/incubator-hivemall/commit/2fa6fb99dd059c2003829e9c455668835e26be24 fixed CI error.

**Comment by myui** [comment]
@takuti CI error is happening. Unit test is failing.

```
[ERROR] /home/travis/build/apache/incubator-hivemall/core/src/test/java/hivemall/ftvec/scaling/L1NormalizationUDFTest.java:[63,41] unreported exception org.apache.hadoop.hive.ql.metadata.HiveException; must be caught or declared to be thrown
```

**Comment by takuti** [comment]
Oh, thanks! Fixed and directly pushed to master.

**Comment by lemire** [comment]
@myui

We had a broken release due to a new layout. This is fixed. Sorry for the problems.

**Comment by myui** [comment]
@lemire 👍  Thank you for creating a great library.

**Comment by lemire** [comment]
@myui There are many of us. 

If you need help, ping us.
